### PR TITLE
[VL] Log more info for shuffle write

### PR DESCRIPTION
**Pull request by zuochunwei**
## What changes were proposed in this pull request?

I proposed this PR to get more detailed information for shuffling writing.

I divided the shuffle writing into several small steps, We can gather the time-consuming for each step after the shuffle writing is completed by setting `VELOX_SHUFFLE_WRITER_LOG_FLAG `.

We can make changes to improve the shuffle writing performance and check the result by this timing info.

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 1]
```diff
@@ -1611,4 +1666,20 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const velox::RowVec
     return shrunken;
   }
 
+  void VeloxShuffleWriter::stat() const {
+#if VELOX_SHUFFLE_WRITER_LOG_FLAG
```
Comment (winningsix) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:1670]: Do we want to report those metrics like [others](https://github.com/oap-project/gluten/blob/da216c134470579f8e7019ab3f69802d94ea0370/cpp/core/shuffle/LocalPartitionWriter.cc#L216)?

Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:1670]: I want to get more information about shuffle writing for tracing performance issues now, You are welcome to report this information as metrics through JNI calling like  doing in `Java_io_glutenproject_vectorized_ShuffleWriterJniWrapper_stop`

**Comment by zhztheplayer** [comment]
@marin-ma Would you like to take a look? Thanks!

File: cpp/velox/shuffle/VeloxShuffleWriter.cc [hunk 2]
```diff
@@ -41,6 +42,8 @@ using namespace facebook::velox;
 
 namespace gluten {
 
+#define VELOX_SHUFFLE_WRITER_LOG_FLAG 1
```
Comment (marin-ma) [inline cpp/velox/shuffle/VeloxShuffleWriter.cc:45]: #define VELOX_SHUFFLE_WRITER_LOG_FLAG 0

File: cpp/core/shuffle/ShuffleWriter.cc [hunk 3]
```diff
@@ -124,6 +126,17 @@ int64_t ShuffleBufferPool::bytesAllocated() const {
...
 
+ShuffleWriter::ShuffleWriter(
+    int32_t numPartitions,
+    std::shared_ptr<PartitionWriterCreator> partitionWriterCreator,
+    const ShuffleWriterOptions& options)
+    : numPartitions_(numPartitions),
+      options_(options),
+      pool_(std::make_shared<ShuffleBufferPool>(options_.memory_pool)),
+      codec_(createArrowIpcCodec(options_.compression_type, options_.codec_backend)) {
+  partitionWriter_ = partitionWriterCreator->make(this).ValueOrDie();
```
Comment (marin-ma) [inline cpp/core/shuffle/ShuffleWriter.cc:137]: Let's avoid directly calling ValueOrDie here. Please refer to Arrow's doc https://arrow.apache.org/docs/cpp/api/support.html#_CPPv4NKR5arrow6Result10ValueOrDieEv

File: cpp/velox/shuffle/VeloxShuffleWriter.cc (hunk 2 above)
Comment (zuochunwei) [reply cpp/velox/shuffle/VeloxShuffleWriter.cc:45]: OK

File: cpp/core/shuffle/ShuffleWriter.cc (hunk 3 above)
Comment (zuochunwei) [reply cpp/core/shuffle/ShuffleWriter.cc:137]: Let me explain why I made this change.

I think partitionWriterCreator is used to create partition writer, and this creating action can be done in the constructor of ShuffleWriter, The partitionWriterCreator's task is done after that and we don't need it anymore.

We need not keep partitionWriterCreator as the member data of VeloxShuffleWriter and be used in the init function of VeloxShuffleWriter.

Comment (marin-ma) [reply cpp/core/shuffle/ShuffleWriter.cc:137]: I agree with your point. But what I concern is that we should avoid calling `ValueOrDie` here.

Comment (marin-ma) [reply cpp/core/shuffle/ShuffleWriter.cc:137]: Maybe we can do more refactors here. Currently we have `LocalPartitionWriterCreator` and `CelebornPartitionWriterCreator`. Should it be better to create one factory for making the `PartitionWriter` instance?

Comment (zuochunwei) [reply cpp/core/shuffle/ShuffleWriter.cc:137]: The XxCreator is taking the role of object factory now.

Comment (zuochunwei) [reply cpp/core/shuffle/ShuffleWriter.cc:137]: using `ARROW_ASSIGN_OR_RAISE` here will cause returning errors of compiling, so if you think it's a problem, let me revert this change here.

Comment (zuochunwei) [reply cpp/core/shuffle/ShuffleWriter.cc]: I revert this change, let's keep the old style.

**Comment by marin-ma** [comment]
@winningsix @FelixYBW Do you have any comments?

**Comment by FelixYBW** [comment]
Let's test the metrics and see if they really count useful values. I'd think many of them actually are very little in pipeline

**Comment by zuochunwei** [comment]
I20230919 09:49:25.320734 106968 VeloxShuffleWriter.cc:373] Velox shuffle writer created, task_attempt_id:9 partitionNum:4000 partitionWriterCreator:N6gluten27LocalPartitionWriterCreatorE partitioning_name:hash buffer_size:4096 compression_type:6 codec_backend:0 compression_mode:0 prefer_evict:0 buffered_write:1 write_eos:0 partition_writer_type:local thread_id:71 offheap_per_task:9641866036
I20230919 09:49:27.930323 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingCompute count: 25, wallNanos: 284883, cpuNanos: 288853 wallNanos-avg:11395 cpuNanos-avg:11554
I20230919 09:49:27.930357 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingBuildPartition count: 25, wallNanos: 487559, cpuNanos: 492563 wallNanos-avg:19502 cpuNanos-avg:19702
I20230919 09:49:27.930362 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingEvictPartition count: 0, wallNanos: 0, cpuNanos: 0
I20230919 09:49:27.930367 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingHasNull count: 25, wallNanos: 13239, cpuNanos: 20772 wallNanos-avg:529 cpuNanos-avg:830
I20230919 09:49:27.930371 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingCalculateBufferSize count: 2118, wallNanos: 229441, cpuNanos: 818565 wallNanos-avg:108 cpuNanos-avg:386
I20230919 09:49:27.930375 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingAllocateBuffer count: 2118, wallNanos: 197147874, cpuNanos: 195445115 wallNanos-avg:93082 cpuNanos-avg:92278
I20230919 09:49:27.930383 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingCreateRbFromBuffer count: 4000, wallNanos: 73552203, cpuNanos: 74162141 wallNanos-avg:18388 cpuNanos-avg:18540
I20230919 09:49:27.930387 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingMakeRB count: 2118, wallNanos: 18510662, cpuNanos: 19069310 wallNanos-avg:8739 cpuNanos-avg:9003
I20230919 09:49:27.930390 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingCacheRB count: 0, wallNanos: 0, cpuNanos: 0
I20230919 09:49:27.930394 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingFlattenRV count: 25, wallNanos: 10806, cpuNanos: 29847 wallNanos-avg:432 cpuNanos-avg:1193
I20230919 09:49:27.930398 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingSplitRV count: 25, wallNanos: 64128193, cpuNanos: 63503576 wallNanos-avg:2565127 cpuNanos-avg:2540143
I20230919 09:49:27.930402 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingIteratePartitions count: 25, wallNanos: 200662788, cpuNanos: 198184359 wallNanos-avg:8026511 cpuNanos-avg:7927374
I20230919 09:49:27.930405 106968 VeloxShuffleWriter.cc:1676] Velox shuffle writer stat:CpuWallTimingStop count: 1, wallNanos: 91156529, cpuNanos: 90465146 wallNanos-avg:91156529 cpuNanos-avg:90465146

**Comment by zuochunwei** [comment]
> Let's test the metrics and see if they really count useful values. I'd think many of them actually are very little in pipeline

image

The result shows `CpuWallTimingAllocateBuffer` is the most time-consuming step.

**Comment by zuochunwei** [comment]
@FelixYBW

**Review by zhztheplayer — APPROVED** [review]

**Pull request merged**